File: tvr-decoder/src/main/java/com/willcurrie/decoders/PopulatedDOLDecoder.java
package com.willcurrie.decoders;

import com.willcurrie.DecodedData;
import com.willcurrie.Decoder;
import com.willcurrie.TagInfo;
import com.willcurrie.TagMetaData;
import com.willcurrie.tlv.ISOUtil;
import com.willcurrie.tlv.Tag;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

public class PopulatedDOLDecoder implements Decoder {
    @Override
    public List<DecodedData> decode(String input, int startIndexInBytes, DecodeSession session) {
        String[] fields = input.split(":");
        String pdol = fields[0];
        String populatedPDOL = fields[1];
        return decode(pdol, populatedPDOL, pdol.length()/2, session);
    }

    public List<DecodedData> decode(String pdol, String populatedPDOL, int startIndexInBytes, DecodeSession session) {
        ArrayList<DecodedData> decoded = new ArrayList<DecodedData>();
        ByteBuffer values = ByteBuffer.wrap(ISOUtil.hex2byte(populatedPDOL));
        List<DOLParser.DOLElement> elements = new DOLParser().parse(ISOUtil.hex2byte(pdol));
        int offset = startIndexInBytes;
        for (DOLParser.DOLElement element : elements) {
            byte[] value = new byte[element.getLength()];
            values.get(value);
            TagMetaData tagMetaData = session.getTagMetaData();
            Tag tag = element.getTag();
            TagInfo tagInfo = tagMetaData.get(tag);
            String valueAsHexString = ISOUtil.hexString(value);
            List<DecodedData> children = tagInfo.getDecoder().decode(valueAsHexString, offset, session);
            DecodedData decodedData = new DecodedData(tag.toString(tagMetaData), tagInfo.decodePrimitiveTlvValue(valueAsHexString), offset, offset + value.length, children);
            decoded.add(decodedData);
            offset += value.length;
        }
        return decoded;
    }

    @Override
    public String validate(String input) {
        String[] fields = input.split(":");
        if (fields.length != 2) {
            return "Put : between the DOL and the populated list";
        }
        return null;
    }

    @Override
    public int getMaxLength() {
        return Integer.MAX_VALUE;
    }
}

File: src/main/java/io/github/binaryfoo/decoders/PopulatedDOLDecoder.kt
package io.github.binaryfoo.decoders

import io.github.binaryfoo.DecodedData
import io.github.binaryfoo.Decoder
import io.github.binaryfoo.TagInfo
import io.github.binaryfoo.TagMetaData
import io.github.binaryfoo.tlv.ISOUtil
import io.github.binaryfoo.tlv.Tag

import java.nio.ByteBuffer
import java.util.ArrayList

public class PopulatedDOLDecoder : Decoder {
    override fun decode(input: String, startIndexInBytes: Int, session: DecodeSession): MutableList<DecodedData> {
        val fields = input.split(":")
        val pdol = fields[0]
        val populatedPDOL = fields[1]
        return decode(pdol, populatedPDOL, pdol.length() / 2, session)
    }

    public fun decode(pdol: String, populatedPDOL: String, startIndexInBytes: Int, session: DecodeSession): MutableList<DecodedData> {
        val decoded = ArrayList<DecodedData>()
        val values = ByteBuffer.wrap(ISOUtil.hex2byte(populatedPDOL))
        val elements = DOLParser().parse(ISOUtil.hex2byte(pdol))
        var offset = startIndexInBytes
        for (element in elements) {
            val value = ByteArray(element.length)
            values.get(value)
            val tagMetaData = session.getTagMetaData()
            val tag = element.tag
            val tagInfo = tagMetaData!!.get(tag)
            val valueAsHexString = ISOUtil.hexString(value)
            val children = tagInfo.decoder.decode(valueAsHexString, offset, session)
            val decodedData = DecodedData.constructed(tag.toString(tagMetaData), tagInfo.decodePrimitiveTlvValue(valueAsHexString), offset, offset + value.size, children)
            decoded.add(decodedData)
            offset += value.size
        }
        return decoded
    }

    override fun validate(input: String?): String? {
        val fields = input?.split(":")?.size ?: 0
        if (fields != 2) {
            return "Put : between the DOL and the populated list"
        }
        return null
    }

    override fun getMaxLength(): Int {
        return Integer.MAX_VALUE
    }
}
